pkg/kn/commands/plugin/path_verifier.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// PathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) []error
+}
+
+// CommandOverrideVerifier verifies that existing kn commands are not overriden
+type CommandOverrideVerifier struct {
+	Root        *cobra.Command
+	SeenPlugins map[string]string
+}
+
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *CommandOverrideVerifier) Verify(path string) []error {
+	if v.Root == nil {
+		return []error{fmt.Errorf("unable to verify path with nil root")}
+	}
+
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
+	if len(cmdPath) > 1 {
+		// the first argument is always "kn" for a plugin binary
+		cmdPath = cmdPath[1:]
+	}
+
+	errors := []error{}
+	isExec, err := isExecutable(path)
+	if err == nil && !isExec {
+		errors = append(errors, fmt.Errorf("warning: %s identified as a kn plugin, but it is not executable", path))
+	} else if err != nil {
+		errors = append(errors, fmt.Errorf("error: unable to identify %s as an executable file: %v", path, err))
+	}
+
+	if existingPath, ok := v.SeenPlugins[binName]; ok {
+		errors = append(errors, fmt.Errorf("warning: %s is overshadowed by a similarly named plugin: %s", path, existingPath))
```

Comment:
Same comments as above - this will just annoy people for what is a valid state. sh doesn't complain about dups on PATH so why should we?  If people really want to help here then let's add some kind of "plugin verify" action that can print all of these kinds of "warnings" for people. Otherwise let's just annoy folks.

Reply:
For me this is where a quiet option as in #182 would help. Once that’s in I will quiet these.

Reply:
the problem with this is that using -q will, IMO, being used all the time and thus hiding everything just not this "valid" state.  I'd like to understand why we're bothering to even do this check at all. There are several conditions that a file had to meet before it'll be consider... why tell people the file doesn't meet them?  Imagine how annoying it would be if each time I can `cat` I got a warning that another `cat` was in my PATH but w/o the +x bit.

Reply:
```
$ kn plugin list
The following compatible plugins are available:

kn-hi
  - warning: kn-hi identified as a kn plugin, but it is not executable
/root/bin/kn-hi
  - warning: /root/bin/kn-hi is overshadowed by a similarly named plugin: kn-hi
```
The first warning isn't right. It's not identified as a plugin because I turned off the +x bit - so showing it in the list is the error, not me having it my PATH.

The exit code on this (which shows no errors just warnings) returns "1", not "0", which means these "warnings" are acting as "errors" - warning do not normally cause a non-zero exit code.

Reply:
Also, 
```
$ kn plugin list
The following compatible plugins are available:

kn-hi
  - warning: kn-hi identified as a kn plugin, but it is not executable
/root/bin/kn-hi
  - warning: /root/bin/kn-hi is overshadowed by a similarly named plugin: kn-hi
Unable read directory "/root/src/k8s.io/kubernetes/_output/dockerized/bin/linux/amd64" from your pluginDir: open /root/src/k8s.io/kubernetes/_output/dockerized/bin/linux/amd64: no such file or directory. Skipping...
error: 2 plugin warnings were found
```
The last line is funny :-)   "error" I found two "warnings".  Aside from that inconsistency in wording, there are actually 3 "warnings" printed, not 2.  IMO none of these are errors or warnings. My system is in a perfectly valid state and so there's no reason to pester me with these messages.